src/client/connection.rs

```diff
@@ -297,30 +331,31 @@ impl Connection {
         let mqtt_state = self.mqtt_state.clone();
         let keep_alive = self.mqttoptions.keep_alive();
         let notification_tx = self.notification_tx.clone();
-
         let network_stream = network_stream
             .map_err(NetworkError::Io)
             .and_then(move |packet| {
                 debug!("Incoming packet = {:?}", packet_info(&packet));
-                let reply = mqtt_state.borrow_mut().handle_incoming_mqtt_packet(packet);
-                future::result(reply)
-            })
-            .and_then(move |(notification, reply)| {
-                handle_notification(notification, &notification_tx);
-                future::ok(reply)
+                let mut mqtt_state = mqtt_state.borrow_mut();
+                mqtt_state.handle_incoming_mqtt_packet(packet)
```

Comment:
This two lines just remove the explicit `future::ok()`. The `Result<>` returned by `handle_incoming_mqtt_packet` can be used directly and coverts because of `IntoFuture` for `Result`. Not sure what you mean by linear and nested her.

Reply:
Oh. Got it. 

What I meant was that method 2 below will keep combinators like `and_then` one after the other instead of one inside the other

```rust
.and_then(move |packet| {
                debug!("Incoming packet = {:?}", packet_info(&packet));
                let mut mqtt_state = mqtt_state.borrow_mut();
                mqtt_state.handle_incoming_mqtt_packet(packet)
                    .and_then(|(notification, reply)| {
                        handle_notification(notification)
                        Ok(reply)
                })
})
.filter(should_forward_packet);
```

vs

```rust
.and_then(move |packet| {
                debug!("Incoming packet = {:?}", packet_info(&packet));
                let mut mqtt_state = mqtt_state.borrow_mut();
                let o = mqtt_state.handle_incoming_mqtt_packet(packet)
                future::result(o)
})
.map(|(notification, reply)| {
        handle_notification(notification)
        reply
})
.filter(should_forward_packet);
```